src/OAuth2.ts

```diff
@@ -0,0 +1,109 @@
+import * as querystring from "querystring";
+import * as rp from "request-promise-native";
+import * as uuid from "uuid/v4";
+import { Logging } from "matrix-appservice-bridge";
+
+import { Main } from "./Main";
+import { BridgedRoom } from "./BridgedRoom";
+import { INTERNAL_ID_LEN } from "./BaseSlackHandler";
+
+const log = Logging.get("OAuth2");
+
+// The full set of OAuth2 scopes we currently require for all functionality
+const REQUIRED_SCOPES = [
+    "team:read",
+    "users:read",
+    "channels:history",
+    "channels:read",
+    "files:write:user",
+    "chat:write:bot",
+    "users:read",
+];
+
+const BOT_SCOPES = [
+    "bot",
+];
+
+export class OAuth2 {
+    private readonly main: Main;
+    private readonly userTokensWaiting: Map<string, string>;
+    private readonly clientId: string;
+    private readonly clientSecret: string;
+    private readonly redirectPrefix: string;
+
+    constructor(opts: {main: Main, client_id: string, client_secret: string, redirect_prefix: string}) {
+        this.main = opts.main;
+        this.userTokensWaiting = new Map(); // token -> userId
+        this.clientId = opts.client_id;
+        this.clientSecret = opts.client_secret;
+        this.redirectPrefix = opts.redirect_prefix;
+    }
+
+    public makeAuthorizeURL(room: string|BridgedRoom, state: string) {
+        const redirectUri = this.makeRedirectURL(room);
+        let scopes = Array.from(REQUIRED_SCOPES);
+        // XXX: Why do we do this?
```

Comment:
great question. In a dedicated commit, please fix it.

Reply:
Reminder to myself to consult halfy from the past.